jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java

```diff
@@ -131,54 +132,80 @@ public void recycle()
         }
     }
 
-    class Bucket
+    public static class Bucket
     {
         private final Deque<ByteBuffer> _queue = new ConcurrentLinkedDeque<>();
         private final ByteBufferPool _pool;
         private final int _capacity;
-        private final AtomicInteger _space;
+        private final int _maxSize;
+        private final AtomicInteger _size;
+        private long _lastUpdate;
 
         public Bucket(ByteBufferPool pool, int capacity, int maxSize)
         {
             _pool = pool;
             _capacity = capacity;
-            _space = maxSize > 0 ? new AtomicInteger(maxSize) : null;
+            _maxSize = maxSize;
+            _size = maxSize > 0 ? new AtomicInteger() : null;
         }
 
+        public ByteBuffer acquire()
+        {
+            ByteBuffer buffer = queuePoll();
+            if (buffer == null)
+                return null;
+            if (_size != null)
+                _size.decrementAndGet();
+            return buffer;
+        }
+
+        /**
+         * @param direct whether to create a direct buffer when none is available
+         * @return a ByteBuffer
+         * @deprecated use {@link #acquire()} instead
+         */
+        @Deprecated
         public ByteBuffer acquire(boolean direct)
         {
             ByteBuffer buffer = queuePoll();
             if (buffer == null)
                 return _pool.newByteBuffer(_capacity, direct);
-            if (_space != null)
-                _space.incrementAndGet();
+            if (_size != null)
+                _size.decrementAndGet();
             return buffer;
         }
 
         public void release(ByteBuffer buffer)
         {
+            _lastUpdate = System.nanoTime();
```

Comment:
Perhaps it is better to update _lastUpdate in acquire. That way we get the last time a buffer of this size was used rather than released.... probably no real difference but feels better to me.

Reply:
Uhm. We do the memory cleanup after a `release()`, so I feel better to update `_lastUpdate` on release rather than acquire.
Also, a bucket may be created and after that its buffers never released, so I would not want to keep it around.